### PR TITLE
Fix empty enum again

### LINKED ISSUE #14: Argument-less enum constructors?
#3 suggests that this is supported. I couldn't find any example usage though - neither in the readme nor in the unit tests.

I tried these two variants:

```haxe
'{ "EnumValue": {} }'
'{ "EnumValue" }'
```

Both produce the same error:

>C:\HaxeToolkit\haxe\lib\tink_json/0,3,1/src/tink/json/macros/GenReader.hx:220: characters 23-29 : TestEnum cannot be called
C:\HaxeToolkit\haxe\lib\tink_json/0,3,1/src/tink/json/macros/Macro.hx:41: lines 41-43 : Defined in this class
source/Main.hx:3: characters 35-56 : Void should be TestEnum
C:\HaxeToolkit\haxe\lib\tink_json/0,3,1/src/tink/json/macros/Macro.hx:41: lines 41-43 : Defined in this class

Am I doing something wrong, or is this a regression?

Here's the full code:

```haxe
class Main {
    public static function main() {
        var o:TestEnum = tink.Json.parse('{ "EnumValue" }');
        trace(o);
    }
}

enum TestEnum {
    EnumValue;
}
```

### PR DESCRIPTION
**Pull request by kevinresol**
Closes #14

### CODE REVIEW

File: src/tink/json/macros/Macro.hx [hunk 1]
```diff
@@ -91,6 +91,9 @@ class Macro {
   }
   
   static public function getRepresentation(t:Type, pos:Position) {
+    
+    if(t.match(TMono(_))) return None;
```
Comment (kevinresol) [inline src/tink/json/macros/Macro.hx]: This is to guard infinite loop if something happens to be `Unknown<0>`.
For example, it happens when doing `tink.Json.stringify(None)`, where the type is `Option<Unknown>`

**Comment by Gama11** [comment]
The `'{ "EnumValue": {} }'` case now works. Shouldn't `'{ "EnumValue" }'` work too though? For that I get:

>Error: Error#422: Expected : at character 14 in { "EnumValue"   ---->  }  <----   @ tink.json.BasicParser.die:237

**Comment by kevinresol** [comment]
`'{ "EnumValue" }'` is invalid json. This lib tries to represent haxe data with json syntax + macro magic. It doesn't modify the json synxtax itself. Without haxe macro, you can still parse it. Only it may not be very meaningful.

**Comment by Gama11** [comment]
Right you are, I was thinking about something like this:

```haxe
class Main {
    public static function main() {
        var o:Struct = tink.Json.parse('{ "e": "EnumValue" }');
        trace(o);
    }
}

typedef Struct = {
    var e:TestEnum;
}

enum TestEnum {
    EnumValue;
}
```

**Comment by kevinresol** [comment]
I think for now the best thing you can do is this
```haxe
enum TestEnum {
    @:json({type:'EnumValue'}) EnumValue;
}
trace(tink.Json.stringify({e:EnumValue})); // {"e":{"type":"EnumValue"}}
```

**Comment by kevinresol** [comment]
> '{ "e": "EnumValue" }'

It is because the reader reads all enum values as object, so it always expects a `{`. (right after the colon in this case) It is a runtime task to determine a enum value to have arguments or not. So I don't think it is possible to support that.

**Comment by Gama11** [comment]
>It is a runtime task to determine a enum value to have arguments or not

How so? You should know whether an enum constructor has arguments or not at compile-time, you know the enum's type after all.

Note that I have very little idea how tink_json works internally. :)

**Comment by kevinresol** [comment]
> It is a runtime task to determine a enum value to have arguments or not

When the type is a Enum, the macro basically generate a piece of code that say: "hey we are expecting a enum, let's read it as an object and get the constructor name from within it". So, the parser will always expect a `{` at that point.
In other words, you can't tell the code to expect `"` instead of `{` just for a no-arg enum value. Because you can't know which enum constructor it is before you parse it.

**Comment by Gama11** [comment]
So it only checks the next character, meaning "expect either `{` or `"EnumValue"`" wouldn't work?

**Comment by kevinresol** [comment]
> So it only checks the next character, meaning "expect either { or "EnumValue"" wouldn't work?

Well, that should work in theory. What I mentioned is the current status quo of the parser.

**Pull request merged**